### PR TITLE
Primary bitstream support for the REST api

### LINKED ISSUE #3349: Restore Primary Bitstream as a capability of DSpace 7
**Is your feature request related to a problem? Please describe.**
Prior versions of DSpace had a field in the Bundle table [1] which provided a link to a bitstream, called 'primary_bitstream_id'[1] This was used to determine what thumbnail to show for a specific item that had multiple bitstreams attached. This would also help for instances of scholarly publications, where you would have the "primary" bitstream as the PDF of the article, and then any supporting bitstreams (images, other media files, software, a data set, etc.)

**Describe the solution you'd like**
There needs to be a way to indicate whether a bitstream is a “primary bitstream” for an item. This should be part of the ingest process, and should be utilized in thumbnail selection for items. This metadata should also be accessible to item templates, so that a designated primary bitstream can be displayed differently on an item page, in order to give it prominence, and to otherwise fit an institution’s use case and site designs. The text extracted from a primary bitstream should also be given a higher relevance ranking in the fulltext search index.

_UPDATE from DevMtg on April 7, 2022:_  All the functionality of "primary bitstreams" still exists in the codebase, but the REST endpoints are missing. We discussed the solution is likely two REST endpoints:
* Submission Side (selecting primary bitstream during submission), perhaps as part of the Upload endpoints: https://github.com/DSpace/RestContract/blob/main/submissionuploads.md
* Bundle Side (updating/selecting primary bitstream while editing): https://github.com/DSpace/RestContract/blob/main/bundles.md
* (NOTE: The individual claiming this ticket could implement both endpoints or just one endpoint...in which case a separate follow-up ticket would be created to implement the other endpoint.)

**Describe alternatives or workarounds you've considered**
Use an earlier version of DSpace, or reorder bitstreams so that the "primary" is listed first (as first bitstream is considered the primary one when no primary is selected)

**Additional context**

* https://wiki.lyrasis.org/display/DSDOC7x/Storage+Layer
* Required by this UI feature request: https://github.com/DSpace/dspace-angular/issues/1578

### PR DESCRIPTION
**Pull request by KevinVdV**
## References
* Fixes #3349
* UI PR: https://github.com/DSpace/dspace-angular/pull/2213

## Description

Rest changes according to the following contract: https://github.com/DSpace/RestContract/pull/222

## Instructions for Reviewers

Take a look at the changes in the contract & try them on the branch.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [X] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [X] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace-api/src/test/java/org/dspace/builder/BundleBuilder.java [hunk 1]
```diff
@@ -47,6 +48,11 @@ public BundleBuilder withName(String name) {
         return this;
     }
 
+    public BundleBuilder withPrimaryBitstream(Bitstream bitstream) {
+        this.primaryBitstream = bitstream;
+        return this;
+    }
+
```
Comment (abollini) [inline dspace-api/src/test/java/org/dspace/builder/BundleBuilder.java:55]: would be better to add an overloaded method `addBitstream` that takes a second parameter `boolean isPrimary` used to set the primary bitstream in the bundle?
in this way the api should look easier to use and there is no risk that someone would set a primary bitstream that is outside the bundle

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/model/BundleRest.java
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/model/BundleRest.java:1]: changes to this class are not required if we implement the REST API according to our general rules using the text/uri-list support

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/BundleConverter.java
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/BundleConverter.java:1]: changes to this class are not required if we implement the REST API according to our general rules using the text/uri-list support

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/BundlePrimaryBitstreamUUIDAddOperation.java
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/BundlePrimaryBitstreamUUIDAddOperation.java:1]: this logic should go in a dedicated controller similar to `org.dspace.app.rest.ItemOwningCollectionUpdateRestController`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/BundlePrimaryBitstreamUUIDRemoveOperation.java
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/BundlePrimaryBitstreamUUIDRemoveOperation.java:1]: this logic should go in a dedicated controller similar to `org.dspace.app.rest.ItemOwningCollectionUpdateRestController`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/BundlePrimaryBitstreamUUIDReplaceOperation.java
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/BundlePrimaryBitstreamUUIDReplaceOperation.java:1]: this logic should go in a dedicated controller similar to `org.dspace.app.rest.ItemOwningCollectionUpdateRestController`

**Review by abollini — CHANGES_REQUESTED** [review]
thanks for this contribution @KevinVdV I have tried to give direction inline to update this PR according to the suggestions in https://github.com/DSpace/DSpace/pull/8798

**Review by tdonohue — CHANGES_REQUESTED** [review]
@KevinVdV : Similar to @abollini , I feel this design doesn't align well with our existing REST API guidelines.  PATCH doesn't seem appropriate, and we should be using PUT instead to create the "primary bitstream" relationship between a Bundle and a Bitstream.  See my review of the contract for details: https://github.com/DSpace/RestContract/pull/222#pullrequestreview-1424862089

If this PR can be refactored to use that approach, that would be best.

**Comment by tdonohue** [comment]
@KevinVdV and @benbosman : Friendly reminder that the Primary Bitstream work (all PRs) needs updating.  This has been sitting for 10 days now, and we'd need it to move more quickly if we are going to have time to review/approve it for 7.6.  Review deadline is this coming Friday (May 26). Merge deadline is June 9.  Thanks!

**Comment by tdonohue** [comment]
Closing, replaced by #8876

**Pull request closed without merging**